File: growthscout-js/src/exceptions.ts
// growthscout-js/src/exceptions.ts
/**
 * Custom exceptions hierarchy for the GrowthScout AI TS/JS SDK.
 */

export class GrowthScoutError extends Error {
  constructor(message: string) {
    super(message);
    this.name = this.constructor.name;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class GrowthScoutNetworkError extends GrowthScoutError {}

export class GrowthScoutAuthenticationError extends GrowthScoutError {}

export class GrowthScoutValidationError extends GrowthScoutError {}

export class GrowthScoutRateLimitError extends GrowthScoutError {
  public retryAfterSeconds: number;

  constructor(message: string, retryAfterSeconds: number = 1.0) {
    super(message);
    this.retryAfterSeconds = retryAfterSeconds;
  }
}

export class GrowthScoutLockConflictError extends GrowthScoutError {}

export class GrowthScoutPayloadTooLargeError extends GrowthScoutError {}

File: growthscout-python/growthscout/exceptions.py
# growthscout-python/growthscout/exceptions.py
"""
Exceptions hierarchy for the GrowthScout AI Python SDK.
"""

class GrowthScoutError(Exception):
    """Base exception class for all SDK errors."""
    pass


class GrowthScoutNetworkError(GrowthScoutError):
    """Raised when request timeouts or host is unreachable."""
    pass


class GrowthScoutAuthenticationError(GrowthScoutError):
    """Raised on HTTP 401 Unauthorized errors."""
    pass


class GrowthScoutValidationError(GrowthScoutError):
    """Raised on HTTP 400 or HTTP 422 validation errors."""
    pass


class GrowthScoutRateLimitError(GrowthScoutError):
    """Raised on HTTP 429 rate limit exceeded. Exposes retry_after duration."""
    def __init__(self, message: str, retry_after: float = 1.0) -> None:
        super().__init__(message)
        self.retry_after = retry_after


class GrowthScoutLockConflictError(GrowthScoutError):
    """Raised on HTTP 409 Conflict errors (session executing/locked)."""
    pass


class GrowthScoutPayloadTooLargeError(GrowthScoutError):
    """Raised on HTTP 413 body size limit violations."""
    pass
